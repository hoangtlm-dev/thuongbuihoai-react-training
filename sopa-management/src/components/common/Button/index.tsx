import React, { memo } from "react";
import "./button.css";

export enum ButtonType {
  default = "",
  primary = "primary",
  secondary = "secondary",
  tertiary = "tertiary",
  btnIconPrimary = "icon-primary",
  btnIconSecondary = "icon-secondary"
};

interface ButtonProps {
  children?: string | JSX.Element;
  type?: ButtonType;
  disable?: boolean;
  onClick?: () => void;
  onSubmit?: () => void;
};

const Button: React.FC<ButtonProps> = ({
  children = "",
  type = ButtonType.default,
  disable = false,
  onClick = () => {},
  onSubmit = () => {},
}) => {
  let className = "btn";
  switch (type) {
    case ButtonType.primary:
      className += ` btn-${ButtonType.primary}`;
      break;
    case ButtonType.secondary:
      className += ` btn-${ButtonType.secondary}`;
      break;
    case ButtonType.tertiary:
      className += ` btn-${ButtonType.tertiary}`;
      break;
    case ButtonType.btnIconPrimary:
      className += ` btn-${ButtonType.btnIconPrimary}`;
      break;
    case ButtonType.btnIconSecondary:
      className += ` btn-${ButtonType.btnIconSecondary}`;
      break;
    default:
      break;
  };

  return (
    <button
      data-testid="button"
      className={`${className} btn-${type}`}
      disabled={disable}
      onClick={onClick}
      onSubmit={onSubmit}
    >{children}</button>
  );
};

export default memo(Button);
