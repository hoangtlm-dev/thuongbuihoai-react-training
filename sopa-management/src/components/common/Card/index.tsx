import { memo } from 'react';
import { Link } from 'react-router-dom';

// Components
import Text,
{
  ThemeType
} from '@common/Text';
import Price,
{
  PriceType
} from '@common/Price';
import Image from '@common/Image';

// Styles
import './card.css';

interface CardProps {
  key?: string;
  name?: string;
  color?: string;
  price?: number;
  href: string;
  src: string;
  alt?: string;
}

const Card: React.FC<CardProps> = ({
  name = '',
  color = '',
  price = 0,
  href,
  src,
  alt = ''
}) => {
  return (
    <div data-testid='card' className='card'>
      <div className='card-item'>
        <figure className='card-image'>
          <Image href={href} url={src} alt={alt} className='card-image-item' />
        </figure>
        <div className='card-info'>
          <Link to={href} className='card-name'>
            {name}
          </Link>
          <div className='card-description'>
            <Text text={color} type={ThemeType.highlightPrimary} />
            <Price value={price} type={PriceType.primary} />
          </div>
        </div>
      </div>
    </div>
  );
};

export default memo(Card);
