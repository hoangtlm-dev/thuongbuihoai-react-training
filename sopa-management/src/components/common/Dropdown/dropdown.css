.custom-dropdown {
  position: relative;
}

.custom-dropdown::after {
  content: '';
  position: absolute;
  left: 335px;
  top: 40%;
  display: inline-block;
  width: 12px;
  height: 8px;
  background-image: url(/images/icons/arrow.svg);
}

.dropdown {
  width: var(--w-medium);
  border-radius: var(--rounded-tiny);
  border: 1px solid var(--border-primary);
  padding: var(--p-base) var(--py-normal);
  -moz-appearance: none;
  -webkit-appearance:none;
  appearance: none;
}
