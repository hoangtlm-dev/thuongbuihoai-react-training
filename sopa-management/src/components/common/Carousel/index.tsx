// Images
import { IMAGE } from '@constants/image';

// Components
import Button,
{
  ButtonType
} from '@common/Button';
import Image from '@common/Image';
import Input from '@common/Input';
import Text,
{
  SizeType
} from '@common/Text';

// Constants
import { CAROUSEL_BULLET } from '@constants/common';

// Styles
import './carousel.css';

const Carousel: React.FC = () => (
  <div data-testid='carousel' className='carousel'>
    <div className='carousel-inner'>
      <Input
        className='carousel-open'
        type='radio'
        id='carousel-1'
        name='carousel'
        hidden
        checked
      />
      <figure className='carousel-item'>
        <Image url={IMAGE.banner} alt={IMAGE.altBanner} />
      </figure>
      <Input
        className='carousel-open'
        type='radio'
        id='carousel-2'
        name='carousel'
        hidden
      />
      <figure className='carousel-item'>
        <Image url={IMAGE.banner2} alt={IMAGE.altBanner} />
      </figure>
      <Input
        className='carousel-open'
        type='radio'
        id='carousel-3'
        name='carousel'
        hidden
      />
      <figure className='carousel-item'>
        <Image url={IMAGE.banner3} alt={IMAGE.altBanner} />
      </figure>
      <ul className='carousel-indicators'>
        {CAROUSEL_BULLET.map((item) => (
          <li key={item.key}>
            <label htmlFor={item.htmlFor} className={item.className} />
          </li>
        ))}
      </ul>
    </div>
    <div className='carousel-info'>
      <Text
        text='Step inside, for comfort and magic await you.'
        type={SizeType.large}
        className='carousel-heading'
      />
      <Button children='Show now' type={ButtonType.primary} />
    </div>
  </div>
);

export default Carousel;
