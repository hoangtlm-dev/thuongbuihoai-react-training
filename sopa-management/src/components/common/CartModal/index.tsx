import { memo } from 'react';

// Images
import { Close } from '../../../../public/images/icons';

// Components
import Text from '../Text';
import Button,
{
  ButtonType
} from '../Button';
import Price,
{
  PriceType
} from '../Price';

// Interfaces
import { Product } from '../../../interfaces/product';

// Helpers
import { totalPrices } from '../../../helpers/common';

// Stores
import { useCartStore } from '../../../stores/cart';

// Styles
import './cartModal.css';
import CartItem from './CartItem';

interface CartModalProps {
  onToggleModal?: () => void;
}
const CartModal: React.FC<CartModalProps> = ({
  onToggleModal = () => {},
}) => {
  const { carts } = useCartStore();

  return (
    <div data-testid='cart-modal' className='overlay'>
      <div className='cart'>
        <div className='cart-header'>
          <Text text='Cart' className='cart-text' />
          <Button
            children={<Close />}
            type={ButtonType.btnIconPrimary}
            onClick={onToggleModal}
          />
        </div>
        <div className='card-body'>
          {carts?.map((cartItem: Product) => (
            <CartItem
              key={cartItem.id}
              cartItem={cartItem}
            />
          ))}
        </div>
        <hr />
        <div className='cart-footer'>
          <Text text='Subtotal' className='cart-text' />
          <Price value={totalPrices(carts)} type={PriceType.tertiary} />
        </div>
      </div>
    </div>
  );
};

export default memo(CartModal);
