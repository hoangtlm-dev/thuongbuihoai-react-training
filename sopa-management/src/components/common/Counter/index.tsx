import { memo } from 'react';

// Components
import Button, { ButtonType } from '@common/Button';
import Input from '../Input';

// Styles
import './counter.css';

interface CounterProps {
  value: number;
  onIncrement: () => void;
  onDecrement: () => void;
}

const Counter = ({
  value,
  onIncrement,
  onDecrement
}: CounterProps) => (
  <div data-testid='counter' className='counter'>
    <Button
      ariaLabel='Decrement'
      type={ButtonType.btnIconPrimary}
      className='counter-minus'
      onClick={onDecrement}
    />
    <Input type='number' classNameInput='counter-input' value={value} />
    <Button
      ariaLabel='Increment'
      type={ButtonType.btnIconPrimary}
      className='counter-plus'
      onClick={onIncrement}
    />
  </div>
);

export default memo(Counter);
