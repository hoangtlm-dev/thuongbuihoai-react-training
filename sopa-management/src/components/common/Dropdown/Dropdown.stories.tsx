import type { Meta, StoryObj } from '@storybook/react';
import Dropdown from '.';
import { SIZE } from '../../../constants/common';

const meta: Meta<typeof Dropdown> = {
  component: Dropdown,
};

export default meta;
type Story = StoryObj<typeof Dropdown>;

export const DropdownSize: Story = {
  args: {
    data: SIZE,
  }
};
