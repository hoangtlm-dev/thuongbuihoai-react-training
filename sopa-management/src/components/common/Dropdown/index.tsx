import React, { memo } from 'react';

// Styles
import './dropdown.css';

type SizeType = {
  key?: string;
  label?: string;
};

interface DropdownProps {
  data?: SizeType[];
}

const Dropdown: React.FC<DropdownProps> = ({ data = [] }) => (
  <div data-testid='dropdown' className='custom-dropdown'>
    <select className='dropdown'>
      {data.map((item: SizeType) => (
        <option key={item.key} value={item.label} className='option'>
          {item.label}
        </option>
      ))}s
    </select>
  </div>
);

export default memo(Dropdown);
