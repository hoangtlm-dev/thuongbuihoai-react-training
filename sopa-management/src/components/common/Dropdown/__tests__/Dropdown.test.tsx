// Libs
import '@testing-library/jest-dom';
import { render } from '@testing-library/react';

// Components
import Dropdown from '..';

// Constants
import { SIZE } from '../../../../constants/common';

describe('Dropdown component', () => {
  test('should render Dropdown component', () => {
    const { getByTestId } = render(<Dropdown data={SIZE} />);
    expect(getByTestId('dropdown')).toBeInTheDocument();
  });

  test('should render Dropdown component', () => {
    const { getByTestId } = render(<Dropdown />);
    expect(getByTestId('dropdown')).toBeInTheDocument();
  });

  test('matches snapshot', () => {
    const { asFragment } = render(<Dropdown data={SIZE} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
