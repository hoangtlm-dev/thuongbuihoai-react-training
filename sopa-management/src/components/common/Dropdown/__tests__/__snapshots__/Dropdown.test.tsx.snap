// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="custom-dropdown"
    data-testid="dropdown"
  >
    <select
      class="dropdown"
    >
      <option
        class="option"
        value="Small"
      >
        Small
      </option>
      <option
        class="option"
        value="Medium"
      >
        Medium
      </option>
      <option
        class="option"
        value="Large"
      >
        Large
      </option>
      s
    </select>
  </div>
</DocumentFragment>
`;
